central/authprovider/datastore/datastore_impl.go

```diff
@@ -55,23 +56,42 @@ func (b *datastoreImpl) UpdateAuthProvider(ctx context.Context, authProvider *st
 	b.lock.Lock()
 	defer b.lock.Unlock()
 
-	exists, err := b.storage.Exists(ctx, authProvider.GetId())
-	if err != nil {
+	if err := b.verifyExistsAndMutable(ctx, authProvider.GetId(), false); err != nil {
 		return err
 	}
-	if !exists {
-		return errox.NotFound.Newf("auth provider with id %q was not found", authProvider.GetId())
-	}
 	return b.storage.Upsert(ctx, authProvider)
 }
 
 // RemoveAuthProvider removes an auth provider from bolt
-func (b *datastoreImpl) RemoveAuthProvider(ctx context.Context, id string) error {
+func (b *datastoreImpl) RemoveAuthProvider(ctx context.Context, deleteReq *v1.DeleteByIDWithForce) error {
 	if ok, err := authProviderSAC.WriteAllowed(ctx); err != nil {
 		return err
 	} else if !ok {
 		return sac.ErrResourceAccessDenied
 	}
 
+	id := deleteReq.GetId()
+	if err := b.verifyExistsAndMutable(ctx, id, deleteReq.GetForce()); err != nil {
+		return err
+	}
 	return b.storage.Delete(ctx, id)
 }
+
+func (b *datastoreImpl) verifyExistsAndMutable(ctx context.Context, id string, force bool) error {
+	provider, exists, err := b.storage.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return errox.NotFound.Newf("auth provider with id %q was not found", id)
+	}
+
+	if provider.GetTraits().GetMutabilityMode() == storage.Traits_ALLOW_MUTATE {
+		return nil
+	}
+
+	if provider.GetTraits().GetMutabilityMode() == storage.Traits_ALLOW_MUTATE_FORCED && !force {
+		return errox.InvalidArgs.Newf("auth provider %q is immutable", id)
+	}
+	return nil
```

Comment:
```suggestion
	switch provider.GetTraits().GetMutabilityMode() {
	case storage.Traits_ALLOW_MUTATE:
		return nil

	case storage.Traits_ALLOW_MUTATE_FORCED:
		if force {
			return nil
		}
		// Tricky - ideally, we would return a message here indicating to the user that they need to pass force=true in order to have the operation succeed. However, that could cause confusion for UI users?
	default:
		// Might be a good place for utils.Should
	}
	return errox.InvalidArgs.Newf("auth provider %q is immutable", id)
```
That way, when we add a new mutability mode but forget to handle it here, we conservatively prevent mutations.

Reply:
➕